<a href="https://colab.research.google.com/github/Dac196/machine-learning/blob/main/Costo_de_vivienda_boston.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

**PARTE 1**

                                            
DataSet costo de vivienda en boston

In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from numpy import genfromtxt
import pandas as pd
dataset = pd.read_csv('https://raw.githubusercontent.com/m-mehdi/tutorials/main/boston_housing.csv', delimiter=',',header=None)
X = dataset.iloc[:,:-1]
y = dataset.iloc[:,-1]

Se importa el conjunto de datos y se crean 2 arreglos, uno con todos los datos excepto la ultima columna que corresponde al valor calculado de la vivienda el segundo solamente guarda el valor de la vivienda. Este sera el que se utilice para comparar el nivel de acierto del script

In [ ]:
dataset.head()

0     1         2    3         4         5         6         7    \
0  0.000000  0.18  0.067815  0.0  0.314815  0.577505  0.641607  0.269203   
1  0.000236  0.00  0.242302  0.0  0.172840  0.547998  0.782698  0.348962   
2  0.000236  0.00  0.242302  0.0  0.172840  0.694386  0.599382  0.348962   
3  0.000293  0.00  0.063050  0.0  0.150206  0.658555  0.441813  0.448545   
4  0.000705  0.00  0.063050  0.0  0.150206  0.687105  0.528321  0.448545   

        8         9    ...       95        96        97        98        99   \
0  0.000000  0.208015  ...  0.059749  0.208015  0.018655  0.082503  0.287234   
1  0.043478  0.104962  ...  0.058064  0.104962  0.021462  0.306021  0.553191   
2  0.043478  0.104962  ...  0.058064  0.103885  0.006661  0.306021  0.547514   
3  0.086957  0.066794  ...  0.043345  0.066412  0.002230  0.421118  0.645222   
4  0.086957  0.066794  ...  0.043345  0.066794  0.006635  0.421118  0.648936   

        100       101       102       103   104  
0  0.025759  1.000000  0.089680  0.008042  24.0  
1  0.113111  1.000000  0.204470  0.041808  21.6  
2  0.035109  0.979580  0.062814  0.004028  34.7  
3  0.021667  0.988585  0.033197  0.001115  33.4  
4  0.064464  1.000000  0.099338  0.009868  36.2  

[5 rows x 105 columns]

In [ ]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Columns: 105 entries, 0 to 104
dtypes: float64(105)
memory usage: 415.2 KB


Se revisa la información general del data set en este caso todos los datos son númericos no hay variables categoricas por lo que se puede continuar con el proceso de entrenamiento del modelo sin la necesidad de excluir columnas o convertir variables catregoricas

In [ ]:
X

0     1         2    3         4         5         6         7    \
0    0.000000  0.18  0.067815  0.0  0.314815  0.577505  0.641607  0.269203   
1    0.000236  0.00  0.242302  0.0  0.172840  0.547998  0.782698  0.348962   
2    0.000236  0.00  0.242302  0.0  0.172840  0.694386  0.599382  0.348962   
3    0.000293  0.00  0.063050  0.0  0.150206  0.658555  0.441813  0.448545   
4    0.000705  0.00  0.063050  0.0  0.150206  0.687105  0.528321  0.448545   
..        ...   ...       ...  ...       ...       ...       ...       ...   
501  0.000633  0.00  0.420455  0.0  0.386831  0.580954  0.681771  0.122671   
502  0.000438  0.00  0.420455  0.0  0.386831  0.490324  0.760041  0.105293   
503  0.000612  0.00  0.420455  0.0  0.386831  0.654340  0.907312  0.094381   
504  0.001161  0.00  0.420455  0.0  0.386831  0.619467  0.889804  0.114514   
505  0.000462  0.00  0.420455  0.0  0.386831  0.473079  0.802266  0.125072   

          8         9    ...       94        95        96        97   \
0    0.000000  0.208015  ...  0.043270  0.059749  0.208015  0.018655   
1    0.043478  0.104962  ...  0.011017  0.058064  0.104962  0.021462   
2    0.043478  0.104962  ...  0.011017  0.058064  0.103885  0.006661   
3    0.086957  0.066794  ...  0.004461  0.043345  0.066412  0.002230   
4    0.086957  0.066794  ...  0.004461  0.043345  0.066794  0.006635   
..        ...       ...  ...       ...       ...       ...       ...   
501  0.000000  0.164122  ...  0.026936  0.146662  0.162090  0.035958   
502  0.000000  0.164122  ...  0.026936  0.146662  0.164122  0.033286   
503  0.000000  0.164122  ...  0.026936  0.146662  0.164122  0.017707   
504  0.000000  0.164122  ...  0.026936  0.146662  0.162694  0.021512   
505  0.000000  0.164122  ...  0.026936  0.146662  0.164122  0.027852   

          98        99        100       101       102       103  
0    0.082503  0.287234  0.025759  1.000000  0.089680  0.008042  
1    0.306021  0.553191  0.113111  1.000000  0.204470  0.041808  
2    0.306021  0.547514  0.035109  0.979580  0.062814  0.004028  
3    0.421118  0.645222  0.021667  0.988585  0.033197  0.001115  
4    0.421118  0.648936  0.064464  1.000000  0.099338  0.009868  
..        ...       ...       ...       ...       ...       ...  
501  0.798551  0.882553  0.195787  0.975392  0.216382  0.048003  
502  0.798551  0.893617  0.181239  1.000000  0.202815  0.041134  
503  0.798551  0.893617  0.096414  1.000000  0.107892  0.011641  
504  0.798551  0.885843  0.117127  0.982677  0.129930  0.017180  
505  0.798551  0.893617  0.151649  1.000000  0.169702  0.028799  

[506 rows x 104 columns]

Se crean los conjuntos de entrenamientos y de testing

En este ejercicio se separa un 30% del dataset para utilizarlo en el entrenamiento del modelo.

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.3, random_state=0)

Se realiza una regresión líneal y se entrena el modelo. Despues se pregunta por los porcentajes de acierto que tuvo en los 2 conjuntos es decir el de entrenamiento y el de testing.

Se puede evidenciar que bajo estas caracteristicas el porcentaje de acierto en el modelo de entrenamiento fue muy alto sin embargo en el de testing el porcenteja es considerablemente más bajo, lo que indica que probablemente se presento un overfitting  y aprendio sobre aprendio del ruido del conjunto de entrenamiento y no es un modelo optimo para predecir resultados con nuevos modelos.

In [ ]:
lr = LinearRegression().fit(X_train, y_train)

print(f"Porcentaje de acierto con el modelo de entrenamiento: {lr.score(X_train, y_train):.2f}")
print(f"Porcentaje de acierto con el modelo de testing: {lr.score(X_test, y_test):.2f}")

Porcentaje de acierto con el modelo de entrenamiento: 0.95
Porcentaje de acierto con el modelo de testing: 0.65


Se utiliza la regularizacion L2 o Ridge con los mismos conjuntos de entrenamiento, se realizan diferentes pruebas de valor alpha, encontrando que con un alpha = 0.1 se mantiene un porcentaje alto en el conjunto de entranmiento y el valor del conjunto de testing tambien incremente su porcentaje de acierto sin embargo la diferencia de aciertos entre los 2 conjuntos de datos sigue siendo alto.

In [ ]:
from sklearn.linear_model import Ridge
ridge = Ridge(alpha=0.1).fit(X_train, y_train)
print(f"Porcentaje de acierto con el modelo de entrenamiento regularizacion Ridge: {ridge.score(X_train, y_train):.2f}")
print(f"Porcentaje de acierto con el modelo de testing regularizacion Ridge: {ridge.score(X_test, y_test):.2f}")

Porcentaje de acierto con el modelo de entrenamiento regularizacion Ridge: 0.93
Porcentaje de acierto con el modelo de testing regularizacion Ridge: 0.80


Se utiliza la regularización L1 o Lasso con los conjuntos de entrenamiento y un valor de alpha = 0.05. Con estas condiciones se encuentra el porcentaje de acierto en el conjunto de testing incrementa sin embargo el porcentaje de acierto del conjunto de entrenamiento disminuye.

In [ ]:
from sklearn.linear_model import Lasso
lasso = Lasso(alpha=0.05).fit(X_train, y_train)
print(f"Porcentaje de acierto con el modelo de entrenamiento regularizacion Lasso: {lasso.score(X_train, y_train):.2f}")
print(f"Porcentaje de acierto con el modelo de testing regularizacion Lasso: {lasso.score(X_test, y_test):.2f}")

Porcentaje de acierto con el modelo de entrenamiento regularizacion Lasso: 0.82
Porcentaje de acierto con el modelo de testing regularizacion Lasso: 0.73


Teniendo en cuenta que Lasso disminuye el peso de algunas variables a 0 se consulta cuantas variables esta utilizando para la predicción del conjunto de datos. En este caso con el valor de alpha asignado se estan utilizando 18 variables de las 103 que tiene el modelo.

Sin embargo si se aumenta el valor de alpha para que incluya más variables los porcentajes de acierto no aumentan significativamente y no ayuda a optimizar el modelo.

In [ ]:
print(f"Number of features: {sum(lasso.coef_ != 0)}")

Number of features: 18
